### PR TITLE
Update Slider widget to support secondaryValue

### LINKED ISSUE #64029: Add secondaryValue to Slider
## Use case

Audio/video players need to display a seek bar, often with a secondary value indicating the buffered position. This allows the user to see how much has downloaded in a different color to how much has played. The Android equivalent of Slider (ProgressBar) supports this via [secondaryProgress](https://developer.android.com/reference/android/widget/ProgressBar#attr_android:secondaryProgress).

There is a flutter plugin that does this called [seekbar](https://pub.dev/packages/seekbar).

Another use case is a progress bar for downloading/installing or downloading/extracting a file where an app may want to show both the download progress and the extraction/installation progress, where one of the progress values is always ahead of the other.

## Proposal

I would propose adding a `secondaryValue` to Slider along with a `secondaryColor`.

One reason to add this into Slider rather than as a separate plugin on pub.dev is that any pub.dev package that does this would have 99% of its code in common with Slider and would not benefit when slider gains new features, and there is also precedent with Android's widget library having this feature built into its Slider equivalent.

### PR DESCRIPTION
**Pull request by xu-baolin**
## Description

**One of the secondary value usages is that Audio/video players need to display a seek bar, often with a secondary value indicating the buffered position.**

Finally, as @clocksmith and @HansMuller suggested and extended `RoundedRectSliderTrackShape` and `RectangularSliderTrackShape` so that developers can easily introduce the second value on the slider.

DEMO:

```dart
import 'package:flutter/material.dart';
import 'dart:async';
import 'dart:math' as math;

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData(
          sliderTheme: SliderThemeData(
//        trackShape: RectangularSliderTrackShape(),
            thumbColor: Colors.red,
//            trackHeight: 20,
          )),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  double value = 0.5;
  double maxValue = 1.0;
  double minValue = 0.0;
  double secondaryVale = 0.6;
  Timer timer;

  @override
  initState() {
    super.initState();
    timer = Timer.periodic(Duration(milliseconds: 100), (timer) {
      setState(() {
        if (secondaryVale == 1.0) {
          secondaryVale = 0.0;
        } else {
          secondaryVale = math.min(1.0, secondaryVale += 0.02);
        }
      });
    });
  }

  @override
  dispose() {
    super.dispose();
    timer.cancel();
  }

  _onChange(newValue) {
    setState(() {
      print('$newValue');
      value = newValue;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        body: Directionality(
//          textDirection: TextDirection.ltr,
          textDirection: TextDirection.rtl,
          child: Center(
            child: SliderTheme(
              data: SliderTheme.of(context).copyWith(
                trackShape: RoundedRectSliderTrackShape(
                  maxValue: maxValue,
                  minValue: minValue,
                  secondaryValue: secondaryVale,
                  secondaryValeTrackColor: Colors.pink,
                ),
              ),
              child: Slider(
                value: value,
                max: maxValue,
                min: minValue,
                onChanged: _onChange,
              ),
            ),
          ),
        ));
  }
}

```

## Related Issues

Fixes #64029 

## Tests

I added the following tests:

See files.

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/material/slider_theme.dart [hunk 1]
```diff
@@ -1536,36 +1558,48 @@ class RectangularSliderTrackShape extends SliderTrackShape with BaseSliderTrackS
...
-      parentBox: parentBox,
-      offset: offset,
-      sliderTheme: sliderTheme,
-      isEnabled: isEnabled,
-      isDiscrete: isDiscrete,
-    );
-
     final Size thumbSize = sliderTheme.thumbShape.getPreferredSize(isEnabled, isDiscrete);
 
-    final Rect leftTrackSegment = Rect.fromLTRB(trackRect.left + trackRect.height / 2, trackRect.top, thumbCenter.dx - thumbSize.width / 2, trackRect.bottom);
```
Comment (xu-baolin) [inline packages/flutter/lib/src/material/slider_theme.dart:1565]: The `leftTrackSegment.left` should not plus `trackRect.height / 2`, it is probably a bug here.
Related commits: https://github.com/flutter/flutter/pull/31564/files#diff-71a1fc327f48d2a30bc5c79f290d5391R1154

Comment (xu-baolin) [reply packages/flutter/lib/src/material/slider_theme.dart:1565]: I changed the `leftTrackSegment.right` from `thumbCenter.dx - thumbSize.width / 2` to `thumbCenter.dx` .
There are two reasons:
1. If the slider's height > thumb's size, there will be a gap 
image
2. The process is different with `RoundedRectSliderTrackShape`

File: packages/flutter/lib/src/material/slider.dart [hunk 2]
```diff
@@ -673,8 +686,10 @@ class _SliderState extends State<Slider> with TickerProviderStateMixin {
     sliderTheme = sliderTheme.copyWith(
       trackHeight: sliderTheme.trackHeight ?? _defaultTrackHeight,
       activeTrackColor: widget.activeColor ?? sliderTheme.activeTrackColor ?? theme.colorScheme.primary,
+      secondaryValueTrackColor: widget.secondaryValueTrackColor ?? sliderTheme.secondaryValueTrackColor ?? theme.colorScheme.primary.withOpacity(0.62),
```
Comment (xu-baolin) [inline packages/flutter/lib/src/material/slider.dart:689]: @HansMuller Hi, I am not sure how to confirm the default value, whom I can ask help? Thanks :)

**Review by xu-baolin — COMMENTED** [review]
This is a draft, I have somethings to be confirmed, and test cases need to be added.

File: packages/flutter/lib/src/material/slider.dart [hunk 3]
```diff
@@ -179,6 +182,7 @@ class Slider extends StatefulWidget {
     Key key,
     @required this.value,
     @required this.onChanged,
+    this.secondaryValue,
```
Comment (xu-baolin) [inline packages/flutter/lib/src/material/slider.dart:185]: I am not sure if need add secondaryValue to both material and cupertino slider.

File: packages/flutter/lib/src/material/slider_theme.dart (hunk 1 above)
Comment (xu-baolin) [reply packages/flutter/lib/src/material/slider_theme.dart:1565]: Track at #64696

**Comment by xu-baolin** [comment]
Ready for review

**Comment by HansMuller** [comment]
CC @clocksmith @JoseAlba

**Comment by willlarche** [comment]
My inclination is that this is too complex to be added without a designer creating the experience. This is not expanding parameters, like our usual PRs, but adding functionality. I believe this is a good case for a community supported package instead. @xu-baolin Please consider forking the slider and sharing it that way.

**Comment by clocksmith** [comment]
This affect should be achievable with the current API by extending SliderTrackShape (https://github.com/flutter/flutter/blob/bbfbf1770c/packages/flutter/lib/src/material/slider_theme.dart#L1122) and painting the track conditionally based on the thumb's location.

**Comment by HansMuller** [comment]
@clocksmith's comment, https://github.com/flutter/flutter/pull/64452#issuecomment-682624664, sounds right to me to. Perhaps we should include an example (or point to one) that demonstrates how to do this.

**Comment by xu-baolin** [comment]
> This affect should be achievable with the current API by extending SliderTrackShape (https://github.com/flutter/flutter/blob/bbfbf1770c/packages/flutter/lib/src/material/slider_theme.dart#L1122) and painting the track conditionally based on the thumb's location.

You mean we should not add the secondary parameters to the slid
er's constructor, just extends the API you mentioned above, right?

**Comment by ryanheise** [comment]
Regarding what the secondary color should be in the default material design theme (light and dark), it appears that the relevant theme file in Android is [progress_horizontal_material.xml](https://android.googlesource.com/platform/frameworks/base/+/master/core/res/res/drawable/progress_horizontal_material.xml). It defines the `secondaryProgress` color as basically `colorAccent` with a different alpha value. Following the references, we can find the alpha values in [colors_material.xml](https://android.googlesource.com/platform/frameworks/base/+/master/core/res/res/values/colors_material.xml) as follows:

```xml
    <item name="disabled_alpha_material_light" format="float" type="dimen">0.26</item>
    <item name="disabled_alpha_material_dark" format="float" type="dimen">0.30</item>
```

**Comment by JoseAlba** [comment]
@xu-baolin 

Here is an example of a way to implement this feature using stacks. 

https://dartpad.dev/d212f07e736edd5e253a3d0df2e32fed

**Comment by xu-baolin** [comment]
> @xu-baolin
> 
> Here is an example of a way to implement this feature using stacks.
> 
> https://dartpad.dev/d212f07e736edd5e253a3d0df2e32fed

Thank you very much.
Through the combination of multiple `slider`s, the effect of `secondaryValue` can indeed be achieved. Here we consider letting the `slider` directly support the `secondaryValue`. This performance is the best.

**Comment by xu-baolin** [comment]
I have implemented the `secondaryValue Slider` by `RectangularSecondarySliderTrackShape` class which extends `SliderTrackShape`, see https://gist.github.com/xu-baolin/ee6fce572a3fff9afe988f2bd754939d or https://dartpad.dev/ee6fce572a3fff9afe988f2bd754939d. I can not open the dart-pad with the gist currently but can copy the code into the pad's code window for running the code.

Talking about my feelings from the perspective of a developer,
1. This implementation is still a little bit complicated, and it may be a little difficult for beginners
2. It’s a little weird because the `secondaryValue` is not a kind of `TrackShape` like `Rectangular` or `RoundedRect`, It is difficult for developers to think of using this to implement this feature by the name

**Comment by xu-baolin** [comment]
@HansMuller @willlarche @clocksmith @ryanheise @JoseAlba 
So let us make a decision on how to support `secondaryValue` for `Slider`, There are several options below,
1) Do not change any code just supply an example (or point to one https://github.com/flutter/flutter/pull/64452#issuecomment-686230659) in the document that demonstrates how to do this
2) Just extending the `RectangularSliderTrackShape` and `RoundedRectSliderTrackShape` API to support that
3) Expanding parameters of `Slider`'s constructor, like the current implementation of this PR
3) Others

On the developer’s side, I prefer the third method, which should be the easiest, and Android is also implemented in this way(https://developer.android.com/reference/android/widget/ProgressBar#attr_android:secondaryProgress). Of course this is just my personal opinion.

**Comment by willlarche** [comment]
Looks like that class from Android is for a linear progress indicator, not slider. Does it still have a knob?

**Comment by xu-baolin** [comment]
> Looks like that class from Android is for a linear progress indicator, not slider. Does it still have a knob?

Yes, you are correct, it does not have the thumb, it called `ProgressBar` and can not found in `https://material.io/components`.
Maybe the most likely widget is `Progress indicators`. From `https://material.io/components/progress-indicators` shows that the Android does not support material `Progress indicators` now. But it does not have a thumb to drag if the usage is Audio/video players seek bar.

In addition, the flutter slider's thumb can be hidden and not be dragged by setting some parameters.

**Comment by HansMuller** [comment]
I agree with https://github.com/flutter/flutter/pull/64452#issuecomment-683699023. Customizing the track to show a secondary value by stacking two sliders, isn't exactly ideal. And we [don't have a UX spec](https://github.com/flutter/flutter/pull/64452#issuecomment-682604740
) for  secondary values yet. I agree that the [example based on extending trackShape](https://github.com/flutter/flutter/pull/64452#issuecomment-683699023) is a bit complicated to include directly in the API doc.  Could we introduce an additional track shape class that makes it easier to define custom tracks that are still essentially rounded rectangles?

**Comment by ryanheise** [comment]
Just to reply to a few comments above:

> Looks like that class from Android is for a linear progress indicator, not slider. Does it still have a knob?

Yes. To clarify, in Android there is a class called SeekBar which is an extension of ProgressBar that adds a draggable thumb, but it inherits all of these properties from ProgressBar by extension.

> I agree with [#64452 (comment)](https://github.com/flutter/flutter/pull/64452#issuecomment-683699023). Customizing the track to show a secondary value by stacking two sliders, isn't exactly ideal. And we [don't have a UX spec](https://github.com/flutter/flutter/pull/64452#issuecomment-682604740) for secondary values yet. I agree that the [example based on extending trackShape](https://github.com/flutter/flutter/pull/64452#issuecomment-683699023) is a bit complicated to include directly in the API doc. Could we introduce an additional track shape class that makes it easier to define custom tracks that are still essentially rounded rectangles?

I think the Android implementation of material design can be used as a practical reference point.

This is a common enough use case that Android's equivalent class has it as a built-in option, so I think if it's not going to be built into Slider, I would at least want an example like the one linked above. The problem with that example is that it's a bit complex to implement for what is a very common use case, but fundamentally I'm not against the idea if it could be made simpler.

Along these lines, it could be useful to be able to add an arbitrary number of overlays to represent secondary values representing different time ranges where the buffer has been loaded. E.g. If the user seeks around and causes discontiguous regions of the buffer to be loaded.

**Comment by xu-baolin** [comment]
> Could we introduce an additional track shape class that makes it easier to define custom tracks that are still essentially rounded rectangles?

How about extending the current `RectangularSliderTrackShape` and  `RoundedRectSliderTrackShape` class to let it support `secondaryValue`?
If a new class is introduced, there will be a lot of duplicate code.

**Comment by HansMuller** [comment]
I apologize for not having time to really dig in here. My initial reaction was that we need to find a way to enable simpler extensions of RectangularSliderTrackShape (or something like it) that would make it easy to introduce a visualization of a secondaryValue - but not just a secondaryValue. So far, I haven't had time make a concrete proposal though.

**Comment by xu-baolin** [comment]
Finally, as above suggested and just extended `RoundedRectSliderTrackShape` and `RectangularSliderTrackShape` so that developers can easily introduce the second value on the slider.
```dart
  Widget build(BuildContext context) {
    return Scaffold(
        body: Directionality(
//          textDirection: TextDirection.ltr,
          textDirection: TextDirection.rtl,
          child: Center(
            child: SliderTheme(
              data: SliderTheme.of(context).copyWith(
                trackShape: RoundedRectSliderTrackShape(
                  maxValue: maxValue,
                  minValue: minValue,
                  secondaryValue: secondaryVale,
                  secondaryValeTrackColor: Colors.pink,
                ),
              ),
              child: Slider(
                value: value,
                max: maxValue,
                min: minValue,
                onChanged: _onChange,
              ),
            ),
          ),
        ));
  }
```

**Comment by xu-baolin** [comment]
@HansMuller Hi,
I'm sorry that this PR lasted a too long time. I think it should be a perfect solution now. Could you review this?

**Comment by Piinks** [comment]
Hi @xu-baolin I am happy to take a look at this if you want to update it. :)

**Comment by Piinks** [comment]
I am going to close this for now since it is not active, but if you would like to continue working on this, feel free to re-open! Thanks as always for your contributions @xu-baolin! :)

**Pull request closed without merging**